pkg/controllers/node/termination/controller.go

```diff
@@ -92,99 +94,148 @@ func (c *Controller) finalize(ctx context.Context, node *corev1.Node) (reconcile
 	// for VolumeAttachments of drain-able Pods to be cleaned up before terminating Node and removing its finalizer.
 	// However, if TerminationGracePeriod is configured for Node, and we are past that period, we will skip waiting.
-	if nodeTerminationTime == nil || c.clock.Now().Before(*nodeTerminationTime) {
-		areVolumesDetached, err := c.ensureVolumesDetached(ctx, node)
-		if err != nil {
-			return reconcile.Result{}, fmt.Errorf("ensuring no volume attachments, %w", err)
-		}
-		if !areVolumesDetached {
-			return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
-		}
-	}
-	nodeClaims, err = nodeutils.GetNodeClaims(ctx, c.kubeClient, node)
+	volumesDetached, err := c.ensureVolumesDetached(ctx, node)
 	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("deleting nodeclaims, %w", err)
+		return reconcile.Result{}, fmt.Errorf("ensuring no volume attachments, %w", err)
 	}
-	for _, nodeClaim := range nodeClaims {
-		isInstanceTerminated, err := termination.EnsureTerminated(ctx, c.kubeClient, nodeClaim, c.cloudProvider)
-		if err != nil {
-			// 404 = the nodeClaim no longer exists
-			if errors.IsNotFound(err) {
-				continue
+	if volumesDetached {
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetTrue(v1.ConditionTypeVolumesDetached); modified {
+			if err := c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+				if errors.IsConflict(err) || errors.IsNotFound(err) {
+					return reconcile.Result{Requeue: true}, nil
+				}
+				return reconcile.Result{}, err
 			}
-			// 409 - The nodeClaim exists, but its status has already been modified
-			if errors.IsConflict(err) {
-				return reconcile.Result{Requeue: true}, nil
+			// We requeue after a patch operation since we want to ensure we read our own writes before any subsequent
+			// operations on the NodeClaim.
+			return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+		}
+	} else if !c.hasTerminationGracePeriodElapsed(nodeTerminationTime) {
+		c.recorder.Publish(terminatorevents.NodeAwaitingVolumeDetachmentEvent(node))
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetFalse(v1.ConditionTypeVolumesDetached, "AwaitingVolumeDetachment", "AwaitingVolumeDetachment"); modified {
+			if err := c.kubeClient.Status().Patch(ctx, nodeClaim, client.MergeFromWithOptions(stored, client.MergeFromWithOptimisticLock{})); err != nil {
+				if errors.IsConflict(err) || errors.IsNotFound(err) {
+					return reconcile.Result{Requeue: true}, nil
+				}
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+	} else {
+		stored := nodeClaim.DeepCopy()
+		if modified := nodeClaim.StatusConditions().SetFalse(v1.ConditionTypeVolumesDetached, "TerminationGracePeriodElapsed", "TerminationGracePeriodElapsed"); modified {
```

Comment:
I could see us setting it to False here since this indicates that we failed to attach the volumes and we had to terminate due to hitting our terminationGracePeriod on the node